File: agent-engine/engine/agent_loop.py
import json
import asyncio
import traceback
import httpx
from engine.session import Session, SessionManager
from engine.skill_loader import SkillLoader
from engine.discussion import DiscussionRunner
from engine.constants import MAX_ITERATIONS, CALLBACK_TIMEOUT, SSE_CHANNEL_BUFFER, DEFAULT_MAX_ROUNDS
from providers.registry import get_provider


class AgentLoop:
    def __init__(self, session_mgr: SessionManager, skill_loader: SkillLoader, go_callback_url: str):
        self.session_mgr = session_mgr
        self.skill_loader = skill_loader
        self.go_callback_url = go_callback_url
        self.discussion_runner = DiscussionRunner(session_mgr, skill_loader, go_callback_url)

    async def run(self, session_id: str):
        session = self.session_mgr.get(session_id)
        if not session:
            return
        try:
            await self._run_inner(session)
        except Exception as e:
            traceback.print_exc()
            session.status = "failed"
            await session.emit({"type": "error", "content": str(e)})
            await session.emit_done()
            await self._notify_go(session)

    async def _notify_go(self, session: Session, advance: bool = False):
        payload = {
            "session_id": session.id,
            "task_id": session.task_id,
            "agent_name": session.agent_name,
            "status": session.status,
            "artifacts": session.artifacts,
            "advance": advance,
        }
        try:
            async with httpx.AsyncClient() as client:
                resp = await client.post(
                    f"{self.go_callback_url}/api/sessions/callback",
                    json=payload,
                    timeout=CALLBACK_TIMEOUT,
                )
                if resp.status_code != 200:
                    print(f"[agent_loop] callback non-200: {resp.status_code}")
        except Exception as e:
            print(f"[agent_loop] callback failed: {e}")

    async def _finish_session(self, session: Session, advance: bool = False):
        if session.status in ("completed", "failed"):
            await session.emit({"type": "session_completed", "agent": session.agent_name})
            await session.emit_done()
        else:
            await session.emit({"type": "session_state_changed", "status": session.status, "agent": session.agent_name})
        await self._notify_go(session, advance=advance)

    async def _run_inner(self, session: Session):
        agent_cfg = self.skill_loader.get_agent_config(session.agent_name)
        if not agent_cfg:
            await session.emit({"type": "error", "content": f"Agent {session.agent_name} not found"})
            await session.emit_done()
            return

        system_prompt = self.skill_loader.get_system_prompt(session.agent_name)
        if session.context:
            system_prompt += f"\n\n## 任务上下文\n{session.context}"

        skills = self.skill_loader.get_skills(session.agent_name)
        tools = [s.to_tool_schema() for s in skills]
        model = self.skill_loader.get_model(session.agent_name)
        provider = get_provider(model)

        await session.add_message("user", f"请开始工作。任务：{session.context}" if session.context else "请开始工作。")

        for _ in range(MAX_ITERATIONS):
            if session.status != "running":
                break

            full_text, tool_calls = await self._stream_llm(provider, session, tools, system_prompt)

            if full_text or tool_calls:
                msg = {"role": "assistant", "content": full_text}
                if tool_calls:
                    msg["tool_calls"] = tool_calls
                await session.add_message("assistant", full_text, tool_calls=tool_calls)

            if not tool_calls:
                if session.status == "running":
                    session.status = "completed"
                await self._finish_session(session)
                return

            for tc in tool_calls:
                await session.emit({"type": "tool_call", "name": tc["name"], "arguments": tc["arguments"], "agent": session.agent_name})

                result = await self._execute_tool(session, tc, skills)
                await session.add_message("tool", result, tool_call_id=tc["id"], name=tc["name"])

                await session.emit({"type": "tool_result", "name": tc["name"], "content": result, "agent": session.agent_name})

                advance = session.status == "completed" and any(a.get("type") == "stage_advance" for a in session.artifacts)
                if session.status in ("completed", "waiting_user"):
                    await self._finish_session(session, advance=advance)
                    return

        session.status = "completed"
        await self._finish_session(session)

    async def _stream_llm(self, provider, session: Session, tools: list, system_prompt: str) -> tuple[str, list[dict]]:
        full_text = ""
        tool_calls = []
        current_tc = None

        async for chunk in provider.stream(messages=session.messages, tools=tools or None, system_prompt=system_prompt):
            if chunk["type"] == "text":
                full_text += chunk["content"]
                await session.emit({"type": "text", "content": chunk["content"], "agent": session.agent_name})
            elif chunk["type"] == "tool_call":
                if current_tc and current_tc not in tool_calls:
                    tool_calls.append(current_tc)
                current_tc = {"id": chunk["id"], "name": chunk["name"], "arguments": chunk["arguments"]}
            elif chunk["type"] == "tool_call_end":
                if current_tc:
                    tool_calls.append(current_tc)
                    current_tc = None
            elif chunk["type"] == "done":
                if current_tc:
                    tool_calls.append(current_tc)
                    current_tc = None

        return full_text, tool_calls

    async def _execute_tool(self, session: Session, tool_call: dict, skills: list) -> str:
        name = tool_call["name"]
        try:
            args = json.loads(tool_call.get("arguments", "{}"))
        except json.JSONDecodeError:
            args = {}

        if name == "advance_stage":
            session.status = "completed"
            session.artifacts.append({"type": "stage_advance", "task_id": session.task_id})
            return "阶段已推进"

        if name == "clarify_user":
            question = args.get("question", "")
            session.status = "waiting_user"
            await session.emit({"type": "clarify_user", "question": question, "agent": session.agent_name})
            await self._notify_go(session)
            user_response = await session.user_input_queue.get()
            await session.add_message("user", user_response)
            session.status = "running"
            return f"用户回复：{user_response}"

        if name == "discuss_with_agent":
            target = args.get("agent_name", "")
            topic = args.get("topic", "")
            max_rounds = args.get("max_rounds", DEFAULT_MAX_ROUNDS)
            discussion_session = self.session_mgr.create(
                agent_name=session.agent_name, task_id=session.task_id, context=topic,
                session_type="discussion", participants=[session.agent_name, target],
                max_rounds=max_rounds, parent_session_id=session.id,
            )
            result = await self.discussion_runner.run(discussion_session.id)
            return f"讨论结论：{result}"

        if name == "write_artifact":
            artifact = {"type": args.get("artifact_type", "document"), "title": args.get("title", ""), "content": args.get("content", "")}
            session.artifacts.append(artifact)
            return "制品已保存"

        skill = self.skill_loader.get_skill(session.agent_name, name)
        if skill:
            try:
                return await skill.execute(**args) if asyncio.iscoroutinefunction(skill.execute) else skill.execute(**args)
            except Exception as e:
                return f"工具执行失败: {e}"

        return f"未知工具: {name}"

    async def run_discussion(self, session_id: str):
        await self.discussion_runner.run(session_id)

File: apps/server/src/engine/agent-loop.ts
import type { Artifact } from "@zcode/shared";
import { LiveSession, sessionManager } from "./session";
import { SkillLoader } from "./skill-loader";
import { getProvider, type LLMProvider } from "./provider";
import * as discussionSvc from "../services/discussion";

const MAX_ITERATIONS = 50;

function builtinTools(): Record<string, unknown>[] {
  return [
    {
      type: "function",
      function: {
        name: "advance_stage",
        description: "确认当前阶段工作完成，推进任务到下一阶段",
        parameters: { type: "object", properties: {}, required: [] },
      },
    },
    {
      type: "function",
      function: {
        name: "write_artifact",
        description: "保存产出物",
        parameters: {
          type: "object",
          properties: {
            artifact_type: { type: "string", description: "制品类型", default: "document" },
            title: { type: "string", description: "制品标题" },
            content: { type: "string", description: "制品内容" },
          },
          required: ["title", "content"],
        },
      },
    },
    {
      type: "function",
      function: {
        name: "discuss_with_agent",
        description: "与另一个 Agent 发起讨论",
        parameters: {
          type: "object",
          properties: {
            agent_name: { type: "string", description: "目标 Agent 名称" },
            topic: { type: "string", description: "讨论主题" },
            max_rounds: { type: "integer", description: "最大讨论轮次", default: 10 },
          },
          required: ["agent_name", "topic"],
        },
      },
    },
    {
      type: "function",
      function: {
        name: "clarify_user",
        description: "向用户提问以澄清问题",
        parameters: {
          type: "object",
          properties: {
            question: { type: "string", description: "要向用户提出的问题" },
          },
          required: ["question"],
        },
      },
    },
  ];
}

export class AgentLoop {
  private skillLoader: SkillLoader;
  private callbackUrl: string;

  constructor(skillLoader: SkillLoader, callbackUrl: string) {
    this.skillLoader = skillLoader;
    this.callbackUrl = callbackUrl;
  }

  async run(sessionId: string): Promise<void> {
    const session = sessionManager.get(sessionId);
    if (!session) return;

    try {
      await this.runInner(session);
    } catch (err: any) {
      session.status = "failed";
      session.emit({ type: "error", content: err.message, agent: session.agentName });
      session.emitDone();
      await this.notifyCallback(session);
    }
  }

  private async runInner(session: LiveSession): Promise<void> {
    const systemPrompt = this.skillLoader.getSystemPrompt(session.agentName);
    const fullPrompt = session.context
      ? systemPrompt + "\n\n## 任务上下文\n" + session.context
      : systemPrompt;

    const skills = this.skillLoader.getSkills(session.agentName);
    const tools = [...skills.map((s) => ({
      type: "function",
      function: { name: s.name, description: s.description, parameters: s.parameters },
    })), ...builtinTools()];

    const model = this.skillLoader.getModel(session.agentName);
    const provider = getProvider(model);

    session.addMessage("user", session.context ? `请开始工作。任务：${session.context}` : "请开始工作。");

    for (let i = 0; i < MAX_ITERATIONS; i++) {
      if (session.status !== "running") break;

      const { fullText, toolCalls } = await this.streamLLM(provider, session, tools, fullPrompt);

      if (fullText || toolCalls.length > 0) {
        session.addMessage("assistant", fullText, toolCalls.length > 0 ? { tool_calls: toolCalls } : undefined);
      }

      if (toolCalls.length === 0) {
        session.status = "completed";
        const autoAdvance = this.skillLoader.getAutoAdvance(session.agentName);
        await this.finishSession(session, autoAdvance);
        return;
      }

      for (const tc of toolCalls) {
        session.emit({ type: "tool_call", name: tc.name, arguments: tc.arguments, agent: session.agentName });
        const result = await this.executeTool(session, tc);
        session.addMessage("tool", result, { tool_call_id: tc.id, name: tc.name });
        session.emit({ type: "tool_result", name: tc.name, content: result, agent: session.agentName });

        const advance = (session.status as string) === "completed" && session.artifacts.some((a) => a.type === "stage_advance");
        if ((session.status as string) === "completed" || (session.status as string) === "waiting_user") {
          await this.finishSession(session, advance);
          return;
        }
      }
    }

    session.status = "completed";
    const autoAdvance = this.skillLoader.getAutoAdvance(session.agentName);
    await this.finishSession(session, autoAdvance);
  }

  private async streamLLM(
    provider: LLMProvider,
    session: LiveSession,
    tools: Record<string, unknown>[],
    systemPrompt: string,
  ): Promise<{ fullText: string; toolCalls: { id: string; name: string; arguments: string }[] }> {
    let fullText = "";
    const toolCalls: { id: string; name: string; arguments: string }[] = [];

    for await (const chunk of provider.stream({
      messages: session.messages,
      tools,
      systemPrompt,
    })) {
      if (chunk.type === "text" && chunk.content) {
        fullText += chunk.content;
        session.emit({ type: "text", content: chunk.content, agent: session.agentName });
      } else if (chunk.type === "tool_call" && chunk.id) {
        toolCalls.push({ id: chunk.id, name: chunk.name || "", arguments: chunk.arguments || "{}" });
      }
    }

    return { fullText, toolCalls };
  }

  private async executeTool(
    session: LiveSession,
    toolCall: { id: string; name: string; arguments: string },
  ): Promise<string> {
    let args: Record<string, unknown>;
    try {
      args = JSON.parse(toolCall.arguments || "{}");
    } catch {
      args = {};
    }

    switch (toolCall.name) {
      case "advance_stage":
        session.status = "completed";
        session.artifacts.push({ type: "stage_advance", content: session.taskId });
        return "阶段已推进";

      case "clarify_user": {
        const question = (args.question as string) || "";
        session.status = "waiting_user";
        session.emit({ type: "clarify_user", question, agent: session.agentName });
        const userResponse = await session.waitForUserInput();
        session.addMessage("user", userResponse);
        session.status = "running";
        return `用户回复：${userResponse}`;
      }

      case "discuss_with_agent": {
        const target = (args.agent_name as string) || "";
        const topic = (args.topic as string) || "";
        if (!target) return "错误：未指定目标 Agent";
        try {
          const board = await discussionSvc.getOrCreateBoard(session.taskId);
          const msg = await discussionSvc.createMessage(board.id, {
            speaker: session.agentName,
            content: topic,
            triggerType: "mention",
            mentions: [target],
          });
          return `已在讨论区 @${target}，消息已发送。等待 ${target} 响应。`;
        } catch (err: any) {
          return `发起讨论失败: ${err.message}`;
        }
      }

      case "write_artifact": {
        const artifact: Artifact = {
          type: (args.artifact_type as string) || "document",
          title: (args.title as string) || "",
          content: (args.content as string) || "",
        };
        session.artifacts.push(artifact);
        return "制品已保存";
      }

      default: {
        const skill = this.skillLoader.getSkill(session.agentName, toolCall.name);
        if (skill) {
          try {
            return await skill.execute(args);
          } catch (err: any) {
            return `工具执行失败: ${err.message}`;
          }
        }
        return `未知工具: ${toolCall.name}`;
      }
    }
  }

  private async finishSession(session: LiveSession, advance: boolean): Promise<void> {
    if (session.status === "completed" || session.status === "failed") {
      session.emit({ type: "session_completed", agent: session.agentName });
    } else {
      session.emit({ type: "session_state_changed", status: session.status, agent: session.agentName });
    }
    session.emitDone();
    await this.notifyCallback(session, advance);
  }

  private async notifyCallback(session: LiveSession, advance = false): Promise<void> {
    const payload = {
      sessionId: session.id,
      taskId: session.taskId,
      agentName: session.agentName,
      status: session.status,
      artifacts: session.artifacts,
      advance,
    };
    try {
      await fetch(`${this.callbackUrl}/api/sessions/callback`, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify(payload),
      });
    } catch {
      // callback failure is non-fatal
    }
  }
}
